### ISSUE TITLE
Search Bar to search through different nodes

### ISSUE THREAD

**Issue by chaitali-pathak** [vasturiano/react-force-graph#210]
Hey, it would be really helpful if a Search bar can be added to search through all different nodes. 
Any ideas on how to do in here?
@vasturiano @micahstubbs @muralimulagalapati

**Comment by ItFlyingStart** [comment]
I'm looking for this feature as well. I'm thinking to adapt the next example [https://github.com/vasturiano/react-force-graph/blob/master/example/click-to-focus/index.html](url)

Is it possible to all graph nodes by using **ref={graphRef}** ?

![image](https://user-images.githubusercontent.com/5562995/103436164-fe0b1f80-4c18-11eb-9dfa-38db270813a5.png)

Find the node by Id, Name, Description etc. And set the focus.

Will this work?

**Comment by micahstubbs** [comment]
this sounds like a nice feature for sure.

It would be fun to make a demo app that does this. I have a feeling that search probably ought to live in the webapp rather than the react-force-graph library, since you have to make decisions about state management and UI presentation. 


hhhhmmmm.  Would be interesting to hear what @vasturiano thinks about this.

**Comment by JonThom** [comment]
@micahstubbs @ItFlyingStart @chaitali-pathak I have implemented such a feature in my app. Something like this should work:

set a reference on the graph:   
  
```
const fgRef = useRef(null);
```
..
```
<ForceGraph2D
    ref={fgRef}
```
...

now access the nodes through `fgRef.current.graphData.nodes`.
this allows you to populate e.g. a dropdown menu.
Now you can, for example, adapt the click-to-focus example (referenced above) so selecting a node zooms to it.

**Comment by ItFlyingStart** [comment]
@JonThom: Thanks for your hint. I will try it out.

**Comment by ItFlyingStart** [comment]
@JonThom: can you show how to access **graphData.nodes**? Because I don't see the property **graphData**. Not sure whether this property is also accessible for **ForceGraph3D**?

![image](https://user-images.githubusercontent.com/5562995/108232639-48058180-7143-11eb-821b-f1b9f390f8b4.png)

`<ForceGraph3D
        ref={fgRef}
        graphData={gData}
...
        nodeThreeObject={formatNodeLayout}
      />`

**Comment by vasturiano** [comment]
@ItFlyingStart in principle you shouldn't need to go through the ref to access the `graphData` object. That's essentially the `gData` structure that you're already passing via the `graphData` prop. So you can simply look in `gData.nodes` for this, just make sure to maintain a reference to that.

As for the search feature, it is certainly doable by scanning through the nodes structure and eventually tie it into zoom or highlight interactions to bring the selected node to focus. 
But, I agree with @micahstubbs that that's probably an application level feature and a bit too specific to have it as a feature in this module. It just provides you with all the building blocks to implement it. :)

**Comment by ItFlyingStart** [comment]
@vasturiano : thanks for your reply. I didn't know that the x y z coordinates will be populated in `gData.nodes` after assigning it to the `graphData` property. Great 👍. Just what I need.

![image](https://user-images.githubusercontent.com/5562995/108494045-5ec4e900-72a7-11eb-987a-48dcd8d356a1.png)

**Comment by vasturiano** [comment]
@ItFlyingStart yes, the component doesn't make a defensive copy of the data structure and simply extends it with the position and velocity attributes.

More info about that in the [d3-force-3d docs](https://github.com/vasturiano/d3-force-3d#simulation_nodes). 👍

**Comment by tomdm83** [comment]
Hello All, 
I have been bracking my head on the above for over a week now . i'm not a java(script) developer 
 but usually i can manage with examples ... . if somebody could help me with a working sample of the above that would be great . i'm even willing to pay for the effort :)

(actually i'm an IBM mainframe developper, i'd love to use this great tool in a website to visulalize mainframe batch flows)
